pkg/template/servicebroker/servicebroker.go

```diff
@@ -1,62 +1,99 @@
+	lister             templatelister.TemplateLister
 	templateNamespaces map[string]struct{}
+	ready              chan struct{}
 }
 
-func NewBroker(restconfig restclient.Config, localSAR client.LocalSubjectAccessReviewsNamespacer, secretsGetter internalversion.SecretsGetter, informers shared.InformerFactory, namespaces []string) *Broker {
+func NewBroker(privrestconfig restclient.Config, privkc kclientset.Interface, infraNamespace string, informer templateinformer.TemplateInformer, namespaces []string) *Broker {
+	informer.Informer().AddIndexers(cache.Indexers{
+		templateapi.TemplateUIDIndex: func(obj interface{}) ([]string, error) {
+			return []string{string(obj.(*templateapi.Template).UID)}, nil
+		}})
+
 	templateNamespaces := map[string]struct{}{}
 	for _, namespace := range namespaces {
 		templateNamespaces[namespace] = struct{}{}
 	}
 
-	return &Broker{
-		secretsGetter:      secretsGetter,
-		localSAR:           localSAR,
-		templateclient:     templateclientset.NewForConfigOrDie(&restconfig).Template(),
-		restconfig:         restconfig,
-		lister:             informers.Templates().Lister(),
+	b := &Broker{
+		informer:           informer.Informer(),
+		lister:             informer.Lister(),
 		templateNamespaces: templateNamespaces,
+		ready:              make(chan struct{}),
 	}
-}
 
-func (b *Broker) getClientsForUsername(username string) (kclientset.Interface, client.Interface, templateclientset.Interface, error) {
-	u := &user.DefaultInfo{Name: username}
+	go func() {
+		// the broker is initialised asynchronously because fetching the service
+		// account token requires the main API server to be running
 
-	oc, err := authclient.NewImpersonatingOpenShiftClient(u, b.restconfig)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		glog.Infof("Template service broker: waiting for authentication token")
 
-	kc, err := authclient.NewImpersonatingKubernetesClientset(u, b.restconfig)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		restconfig, oc, kc, _, err := serviceaccounts.Clients(
+			privrestconfig,
+			&serviceaccounts.ClientLookupTokenRetriever{Client: privkc},
+			infraNamespace,
+			bootstrappolicy.InfraTemplateServiceBrokerServiceAccountName,
+		)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("Template service broker: failed to initialize: %v", err))
+			return
+		}
 
-	impersonatingConfig := authclient.NewImpersonatingConfig(u, b.restconfig)
-	templateclient, err := templateclientset.NewForConfig(&impersonatingConfig)
-	if err != nil {
-		return nil, nil, nil, err
-	}
+		b.oc = oc
+		b.kc = kc
+		b.templateclient = templateclientset.NewForConfigOrDie(restconfig).Template()
+
+		glog.Infof("Template service broker: waiting for informer sync")
 
-	return kc, oc, templateclient, nil
+		for !b.informer.HasSynced() {
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		glog.Infof("Template service broker: ready")
+
+		close(b.ready)
+	}()
+
+	return b
+}
+
+func (b *Broker) waitForReady() error {
```

Comment:
I expected this to be called from a filter function in your handler chain so that we could easily make sure the entire handler was covered.